### ISSUE TITLE
Runtime error after minifying React.js library with babili-webpack-plugin

### ISSUE THREAD

**Issue by filmic** [babel/babili#407]
Babili seems to remove property `ReactComponent.prototype.isReactComponent` during minification of `ReactComponent` which crashes React component rendering.

Source of `node_modules/react/lib/ReactComponent.js`:

Non-minified:
```
/***/ "NT9266":
/***/ (function(module, exports, __webpack_require__) {

"use strict";
/**
 * Copyright 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 */



var _prodInvariant = __webpack_require__("QHtMVm");

var ReactNoopUpdateQueue = __webpack_require__("ThfBjY");

var canDefineProperty = __webpack_require__("W6iNs+");
var emptyObject = __webpack_require__("zEobd7");
var invariant = __webpack_require__("nTc3MK");
var warning = __webpack_require__("do5FhS");

/**
 * Base class helpers for the updating state of a component.
 */
function ReactComponent(props, context, updater) {
  this.props = props;
  this.context = context;
  this.refs = emptyObject;
  // We initialize the default updater but the real one gets injected by the
  // renderer.
  this.updater = updater || ReactNoopUpdateQueue;
}

ReactComponent.prototype.isReactComponent = {};

/**
 * Sets a subset of the state. Always use this to mutate
 * state. You should treat `this.state` as immutable.
 *
 * There is no guarantee that `this.state` will be immediately updated, so
 * accessing `this.state` after calling this method may return the old value.
 *
 * There is no guarantee that calls to `setState` will run synchronously,
 * as they may eventually be batched together.  You can provide an optional
 * callback that will be executed when the call to setState is actually
 * completed.
 *
 * When a function is provided to setState, it will be called at some point in
 * the future (not synchronously). It will be called with the up to date
 * component arguments (state, props, context). These values can be different
 * from this.* because your function may be called after receiveProps but before
 * shouldComponentUpdate, and this new state, props, and context will not yet be
 * assigned to this.
 *
 * @param {object|function} partialState Next partial state or function to
 *        produce next partial state to be merged with current state.
 * @param {?function} callback Called after state is updated.
 * @final
 * @protected
 */
ReactComponent.prototype.setState = function (partialState, callback) {
  !(typeof partialState === 'object' || typeof partialState === 'function' || partialState == null) ?  true ? invariant(false, 'setState(...): takes an object of state variables to update or a function which returns an object of state variables.') : _prodInvariant('85') : void 0;
  this.updater.enqueueSetState(this, partialState);
  if (callback) {
    this.updater.enqueueCallback(this, callback, 'setState');
  }
};

/**
 * Forces an update. This should only be invoked when it is known with
 * certainty that we are **not** in a DOM transaction.
 *
 * You may want to call this when you know that some deeper aspect of the
 * component's state has changed but `setState` was not called.
 *
 * This will not invoke `shouldComponentUpdate`, but it will invoke
 * `componentWillUpdate` and `componentDidUpdate`.
 *
 * @param {?function} callback Called after update is complete.
 * @final
 * @protected
 */
ReactComponent.prototype.forceUpdate = function (callback) {
  this.updater.enqueueForceUpdate(this);
  if (callback) {
    this.updater.enqueueCallback(this, callback, 'forceUpdate');
  }
};

/**
 * Deprecated APIs. These APIs used to exist on classic React classes but since
 * we would like to deprecate them, we're not going to move them over to this
 * modern base class. Instead, we define a getter that warns if it's accessed.
 */
if (false) {
  var deprecatedAPIs = {
    isMounted: ['isMounted', 'Instead, make sure to clean up subscriptions and pending requests in ' + 'componentWillUnmount to prevent memory leaks.'],
    replaceState: ['replaceState', 'Refactor your code to use setState instead (see ' + 'https://github.com/facebook/react/issues/3236).']
  };
  var defineDeprecationWarning = function (methodName, info) {
    if (canDefineProperty) {
      Object.defineProperty(ReactComponent.prototype, methodName, {
        get: function () {
           true ? warning(false, '%s(...) is deprecated in plain JavaScript React classes. %s', info[0], info[1]) : void 0;
          return undefined;
        }
      });
    }
  };
  for (var fnName in deprecatedAPIs) {
    if (deprecatedAPIs.hasOwnProperty(fnName)) {
      defineDeprecationWarning(fnName, deprecatedAPIs[fnName]);
    }
  }
}

module.exports = ReactComponent;

/***/ }),
```

Minified:
```
    NT9266: function(t, o, r) {
        "use strict";
        function p(D, M, w) {
            this.props = D,
            this.context = M,
            this.refs = C,
            this.updater = w || h
        }
        var u = r("QHtMVm")
          , h = r("ThfBjY")
          , g = r("W6iNs+")
          , C = r("zEobd7")
          , v = r("nTc3MK")
          , N = r("do5FhS");
        var T, I, P;
        t.exports = p
    },
```

Rendered component: 
```
import React, { PropTypes } from 'react';

export default class HelloWorld extends React.Component {
  render() {
    return (
      <p>Hello World!</p>
    );
  }
}
```

Babel settings:
```
{
  "presets": ["stage-2", "react"]
}
```

Error:
```
Uncaught TypeError: Class constructor f cannot be invoked without 'new'
    at M._constructComponentWithoutOwner (react-libs.js:formatted:4171)
    at M._constructComponent (react-libs.js:formatted:4167)
    at M.mountComponent (react-libs.js:formatted:4146)
    at Object.mountComponent (react-libs.js:formatted:2878)
    at M.performInitialMount (react-libs.js:formatted:4199)
    at M.mountComponent (react-libs.js:formatted:4162)
    at Object.mountComponent (react-libs.js:formatted:2878)
    at g (react-libs.js:formatted:2684)
    at perform (react-libs.js:formatted:2292)
    at C (react-libs.js:formatted:2691)
```

Webpack uses `"babili-webpack-plugin": "^0.0.9"`

**Comment by vigneshshanmugam** [comment]
I did try out the same code and presets as you have mentioned here, It works totally fine

check the webpack config 
```js
const path = require('path');
const Babili = require('babili-webpack-plugin');
const webpack = require('webpack');

module.exports = {
  entry: path.join(__dirname, 'app.js'),
  output: {
    path: "./",
    filename: "app.bundle.js"
  },
  module: {
    loaders: [
      {
        test: /\.js$/,
        loader: 'babel-loader',
        query: {
          presets: ['stage-2', 'react']
        }
      }
    ]
  },
  plugins: [
    new webpack.DefinePlugin({
          'process.env':{
              'NODE_ENV': JSON.stringify('production')
          }
      }),
    new Babili()
  ]
}
```

**Comment by filmic** [comment]
Hi @vigneshshanmugam 

It's not exactly the same setup I have (I use `webpack.optimize.CommonsChunkPlugin` to split out react libs into vendor chunk and `webpack.HashedModuleIdsPlugin` for modules names).

You can easily replicate the issue using Babel REPL.

Here is [Babel REPL link with code](http://babeljs.io/repl/#?babili=true&evaluate=false&lineWrap=false&presets=react%2Cstage-2&targets=&browsers=&builtIns=false&experimental=false&loose=false&spec=false&code=_webpackJsonp(%5B28%2C%2032%5D%2C%20%7B%0A%20%20%2F***%2F%20%22NT9266%22%3A%0A%2F***%2F%20(function(module%2C%20exports%2C%20__webpack_require__)%20%7B%0A%0A%22use%20strict%22%3B%0A%2F**%0A%20*%20Copyright%202013-present%2C%20Facebook%2C%20Inc.%0A%20*%20All%20rights%20reserved.%0A%20*%0A%20*%20This%20source%20code%20is%20licensed%20under%20the%20BSD-style%20license%20found%20in%20the%0A%20*%20LICENSE%20file%20in%20the%20root%20directory%20of%20this%20source%20tree.%20An%20additional%20grant%0A%20*%20of%20patent%20rights%20can%20be%20found%20in%20the%20PATENTS%20file%20in%20the%20same%20directory.%0A%20*%0A%20*%2F%0A%0A%0A%0Avar%20_prodInvariant%20%3D%20__webpack_require__(%22QHtMVm%22)%3B%0A%0Avar%20ReactNoopUpdateQueue%20%3D%20__webpack_require__(%22ThfBjY%22)%3B%0A%0Avar%20canDefineProperty%20%3D%20__webpack_require__(%22W6iNs%2B%22)%3B%0Avar%20emptyObject%20%3D%20__webpack_require__(%22zEobd7%22)%3B%0Avar%20invariant%20%3D%20__webpack_require__(%22nTc3MK%22)%3B%0Avar%20warning%20%3D%20__webpack_require__(%22do5FhS%22)%3B%0A%0A%2F**%0A%20*%20Base%20class%20helpers%20for%20the%20updating%20state%20of%20a%20component.%0A%20*%2F%0Afunction%20ReactComponent(props%2C%20context%2C%20updater)%20%7B%0A%20%20this.props%20%3D%20props%3B%0A%20%20this.context%20%3D%20context%3B%0A%20%20this.refs%20%3D%20emptyObject%3B%0A%20%20%2F%2F%20We%20initialize%20the%20default%20updater%20but%20the%20real%20one%20gets%20injected%20by%20the%0A%20%20%2F%2F%20renderer.%0A%20%20this.updater%20%3D%20updater%20%7C%7C%20ReactNoopUpdateQueue%3B%0A%7D%0A%0AReactComponent.prototype.isReactComponent%20%3D%20%7B%7D%3B%0A%0A%2F**%0A%20*%20Sets%20a%20subset%20of%20the%20state.%20Always%20use%20this%20to%20mutate%0A%20*%20state.%20You%20should%20treat%20%60this.state%60%20as%20immutable.%0A%20*%0A%20*%20There%20is%20no%20guarantee%20that%20%60this.state%60%20will%20be%20immediately%20updated%2C%20so%0A%20*%20accessing%20%60this.state%60%20after%20calling%20this%20method%20may%20return%20the%20old%20value.%0A%20*%0A%20*%20There%20is%20no%20guarantee%20that%20calls%20to%20%60setState%60%20will%20run%20synchronously%2C%0A%20*%20as%20they%20may%20eventually%20be%20batched%20together.%20%20You%20can%20provide%20an%20optional%0A%20*%20callback%20that%20will%20be%20executed%20when%20the%20call%20to%20setState%20is%20actually%0A%20*%20completed.%0A%20*%0A%20*%20When%20a%20function%20is%20provided%20to%20setState%2C%20it%20will%20be%20called%20at%20some%20point%20in%0A%20*%20the%20future%20(not%20synchronously).%20It%20will%20be%20called%20with%20the%20up%20to%20date%0A%20*%20component%20arguments%20(state%2C%20props%2C%20context).%20These%20values%20can%20be%20different%0A%20*%20from%20this.*%20because%20your%20function%20may%20be%20called%20after%20receiveProps%20but%20before%0A%20*%20shouldComponentUpdate%2C%20and%20this%20new%20state%2C%20props%2C%20and%20context%20will%20not%20yet%20be%0A%20*%20assigned%20to%20this.%0A%20*%0A%20*%20%40param%20%7Bobject%7Cfunction%7D%20partialState%20Next%20partial%20state%20or%20function%20to%0A%20*%20%20%20%20%20%20%20%20produce%20next%20partial%20state%20to%20be%20merged%20with%20current%20state.%0A%20*%20%40param%20%7B%3Ffunction%7D%20callback%20Called%20after%20state%20is%20updated.%0A%20*%20%40final%0A%20*%20%40protected%0A%20*%2F%0AReactComponent.prototype.setState%20%3D%20function%20(partialState%2C%20callback)%20%7B%0A%20%20!(typeof%20partialState%20%3D%3D%3D%20'object'%20%7C%7C%20typeof%20partialState%20%3D%3D%3D%20'function'%20%7C%7C%20partialState%20%3D%3D%20null)%20%3F%20%20true%20%3F%20invariant(false%2C%20'setState(...)%3A%20takes%20an%20object%20of%20state%20variables%20to%20update%20or%20a%20function%20which%20returns%20an%20object%20of%20state%20variables.')%20%3A%20_prodInvariant('85')%20%3A%20void%200%3B%0A%20%20this.updater.enqueueSetState(this%2C%20partialState)%3B%0A%20%20if%20(callback)%20%7B%0A%20%20%20%20this.updater.enqueueCallback(this%2C%20callback%2C%20'setState')%3B%0A%20%20%7D%0A%7D%3B%0A%0A%2F**%0A%20*%20Forces%20an%20update.%20This%20should%20only%20be%20invoked%20when%20it%20is%20known%20with%0A%20*%20certainty%20that%20we%20are%20**not**%20in%20a%20DOM%20transaction.%0A%20*%0A%20*%20You%20may%20want%20to%20call%20this%20when%20you%20know%20that%20some%20deeper%20aspect%20of%20the%0A%20*%20component's%20state%20has%20changed%20but%20%60setState%60%20was%20not%20called.%0A%20*%0A%20*%20This%20will%20not%20invoke%20%60shouldComponentUpdate%60%2C%20but%20it%20will%20invoke%0A%20*%20%60componentWillUpdate%60%20and%20%60componentDidUpdate%60.%0A%20*%0A%20*%20%40param%20%7B%3Ffunction%7D%20callback%20Called%20after%20update%20is%20complete.%0A%20*%20%40final%0A%20*%20%40protected%0A%20*%2F%0AReactComponent.prototype.forceUpdate%20%3D%20function%20(callback)%20%7B%0A%20%20this.updater.enqueueForceUpdate(this)%3B%0A%20%20if%20(callback)%20%7B%0A%20%20%20%20this.updater.enqueueCallback(this%2C%20callback%2C%20'forceUpdate')%3B%0A%20%20%7D%0A%7D%3B%0A%0A%2F**%0A%20*%20Deprecated%20APIs.%20These%20APIs%20used%20to%20exist%20on%20classic%20React%20classes%20but%20since%0A%20*%20we%20would%20like%20to%20deprecate%20them%2C%20we're%20not%20going%20to%20move%20them%20over%20to%20this%0A%20*%20modern%20base%20class.%20Instead%2C%20we%20define%20a%20getter%20that%20warns%20if%20it's%20accessed.%0A%20*%2F%0Aif%20(true)%20%7B%0A%20%20var%20deprecatedAPIs%20%3D%20%7B%0A%20%20%20%20isMounted%3A%20%5B'isMounted'%2C%20'Instead%2C%20make%20sure%20to%20clean%20up%20subscriptions%20and%20pending%20requests%20in%20'%20%2B%20'componentWillUnmount%20to%20prevent%20memory%20leaks.'%5D%2C%0A%20%20%20%20replaceState%3A%20%5B'replaceState'%2C%20'Refactor%20your%20code%20to%20use%20setState%20instead%20(see%20'%20%2B%20'https%3A%2F%2Fgithub.com%2Ffacebook%2Freact%2Fissues%2F3236).'%5D%0A%20%20%7D%3B%0A%20%20var%20defineDeprecationWarning%20%3D%20function%20(methodName%2C%20info)%20%7B%0A%20%20%20%20if%20(canDefineProperty)%20%7B%0A%20%20%20%20%20%20Object.defineProperty(ReactComponent.prototype%2C%20methodName%2C%20%7B%0A%20%20%20%20%20%20%20%20get%3A%20function%20()%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20true%20%3F%20warning(false%2C%20'%25s(...)%20is%20deprecated%20in%20plain%20JavaScript%20React%20classes.%20%25s'%2C%20info%5B0%5D%2C%20info%5B1%5D)%20%3A%20void%200%3B%0A%20%20%20%20%20%20%20%20%20%20return%20undefined%3B%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D)%3B%0A%20%20%20%20%7D%0A%20%20%7D%3B%0A%20%20for%20(var%20fnName%20in%20deprecatedAPIs)%20%7B%0A%20%20%20%20if%20(deprecatedAPIs.hasOwnProperty(fnName))%20%7B%0A%20%20%20%20%20%20defineDeprecationWarning(fnName%2C%20deprecatedAPIs%5BfnName%5D)%3B%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A%0Amodule.exports%20%3D%20ReactComponent%3B%0A%0A%2F***%2F%20%7D)%0A%7D)) from issue description.

Note that all properties on `ReactComponent.prototype` are missing from minified code.

If I disable dead code removal in Babili plugin (`deadcode: false`) the code is minified correctly.

**Comment by vigneshshanmugam** [comment]
@filmic This `prototype` issue has already been fixed in master. Please wait for the new version!

**Comment by filmic** [comment]
Great! Awaiting new release then :)

**Comment by filmic** [comment]
Fixed in babili@0.0.11
